pykeepass/kdbx_parsing/factorinfo.py

```diff
@@ -0,0 +1,490 @@
+        elif uuid == FACTOR_TYPE_KEY_FILE:
+            return KeyFileFactor.decode(ret, factor_element)
+        elif uuid == FACTOR_TYPE_EMPTY:
+            return NopFactor.decode(ret, factor_element)
+
+        return ret
+
+    def wrap_key_part(self, user_supplied_info, key_part, wrapping_key = None):
+        factor_name = factor_types_to_names.get(self.uuid, self.uuid)
+        if wrapping_key is None:
+            wrapping_key = self.get_wrapping_key(user_supplied_info=user_supplied_info)
+
+        encrypted_key = None
+        if self.key_type == FACTOR_ALG_AES_CBC:
+            cipher = AES.new(wrapping_key, AES.MODE_CBC, iv=self.key_salt)
+            encrypted_key = cipher.encrypt(key_part)
+
+        if encrypted_key is None:
+            raise NotImplementedError(
+                "Cannot wrap a key part for unknown alg {} on factor type {}".format(self.key_type, factor_name)
+            )
+
+        validation_out = None
+        if self.group.validation_type == FACTOR_VALIDATE_HMAC_SHA512:
+            validation_out = hmac.new(key_part, self.group.validation_in, 'SHA-512').digest()
+        else:
+            raise NotImplementedError(
+                "Cannot verify a key part for unknown alg {} on factor type {}".format(self.group.verify_type,
+                                                                                       factor_name)
+            )
+
+        return encrypted_key, validation_out
+
+    def unwrap_key_part(self, user_supplied_info, unwrapping_key = None):
+        factor_name = factor_types_to_names.get(self.uuid, self.uuid)
+        if unwrapping_key is None:
+            unwrapping_key = self.get_wrapping_key(user_supplied_info=user_supplied_info)
+            if unwrapping_key is None:
+                raise CredentialsError("Could not get key part for factor type {}".format(factor_name))
+
+        decrypted_key = None
+
+        if self.wrapped_key_part is None:
+            # Saving for the first time - new key part
+            cipher = AES.new(unwrapping_key, AES.MODE_CBC, iv=self.key_salt)
+            new_generated_key = random.randbytes(32)
+            self.wrapped_key_part = cipher.encrypt(new_generated_key)
+
+        if self.key_type == FACTOR_ALG_AES_CBC:
+            # Salt forms the AES-CBC IV
+            cipher = AES.new(unwrapping_key, AES.MODE_CBC, iv=self.key_salt)
+
+            # Decrypt wrapped key part
+            decrypted_key = cipher.decrypt(self.wrapped_key_part)
+
+        if decrypted_key is None:
+            raise NotImplementedError(
+                "Cannot unwrap a key part for unknown alg {} on factor type {}".format(self.key_type, factor_name)
+            )
+
+        digest = None
+        if self.group.validation_type == FACTOR_VALIDATE_HMAC_SHA512:
+            digest = hmac.new(decrypted_key, self.group.validation_in, 'SHA-512').digest()
+        else:
+            # Can't verify, we don't know how
+            pass
```

Comment:
In the event it's wrong we'll just end up failing to open the database later, as the constructed composite key won't pass the header checksum.